src/algebra/algebra/subalgebra.lean

```diff
@@ -868,6 +868,28 @@ instance to_algebra {R A : Type*} [comm_semiring R] [comm_semiring A] [semiring
   [algebra R A] [algebra A α] (S : subalgebra R A) : algebra S α :=
 algebra.of_subsemiring S.to_subsemiring
 
+@[simp] lemma srange_algebra_map {R A : Type*} [comm_semiring R] [comm_semiring A]
+  [algebra R A] (S : subalgebra R A) :
+  (algebra_map S A).srange = S.to_subsemiring :=
+begin
+  ext x,
+  rw [ring_hom.mem_srange, subalgebra.mem_to_subsemiring],
+  split,
+  { rintro ⟨⟨_, hx⟩, rfl⟩, exact hx },
+  { exact λ hx, ⟨⟨x, hx⟩, rfl⟩ }
+end
```

Comment:
This proof breaks down into a bunch of smaller missing-but-useful lemmas:
```lean
-- this belongs near `algebra.id.map_eq_self`
@[simp] lemma _root_.algebra.id.map_eq_id : algebra_map R R = ring_hom.id _ := rfl

lemma algebra_map_eq {R A : Type*} [comm_semiring R] [comm_semiring A] [semiring α]
  [algebra R A] [algebra A α] (S : subalgebra R A) :
    algebra_map S α = (algebra_map A α).comp S.val := rfl

-- this belongs near `subalgebra.srange_val`
@[simp] lemma srange_val_coe_ring_hom : (S.val : S →+* A).srange = S.to_subsemiring :=
set_like.coe_injective subtype.range_val

@[simp] lemma srange_algebra_map {R A : Type*} [comm_semiring R] [comm_semiring A]
  [algebra R A] (S : subalgebra R A) :
  (algebra_map S A).srange = S.to_subsemiring :=
by rw [algebra_map_eq, algebra.id.map_eq_id, ring_hom.id_comp, srange_val_coe_ring_hom]
```

Reply:
I ended up defining `val_coe_ring_hom` the other way around as a `simp` lemma, `S.val` seems like a decent `simp` normal form.